### PR TITLE
Remove Atomic's _owned methods

### LINKED ISSUE #30: Removing *_owned methods
Have a look at the following methods on `Atomic<T>` we currently implement:

<details>
  <summary>Click to expand</summary>

```rust
impl<T> Atomic<T> {
    pub fn store(&self, new: Ptr<T>, ord: Ordering) {
        self.data.store(new.data, ord);
    }

    pub fn store_owned(&self, new: Owned<T>, ord: Ordering) {
        let data = new.data;
        mem::forget(new);
        self.data.store(data, ord);
    }

    pub fn swap<'scope>(&self, new: Ptr<T>, ord: Ordering, _: &'scope Scope) -> Ptr<'scope, T> {
        Ptr::from_data(self.data.swap(new.data, ord))
    }

    pub fn compare_and_set<'scope, O>(
        &self,
        current: Ptr<T>,
        new: Ptr<T>,
        ord: O,
        _: &'scope Scope,
    ) -> Result<(), Ptr<'scope, T>>
    where
        O: CompareAndSetOrdering,
    {
        match self.data.compare_exchange(
            current.data,
            new.data,
            ord.success(),
            ord.failure(),
        ) {
            Ok(_) => Ok(()),
            Err(previous) => Err(Ptr::from_data(previous)),
        }
    }

    pub fn compare_and_set_owned<'scope, O>(
        &self,
        current: Ptr<T>,
        new: Owned<T>,
        ord: O,
        _: &'scope Scope,
    ) -> Result<Ptr<'scope, T>, (Ptr<'scope, T>, Owned<T>)>
    where
        O: CompareAndSetOrdering,
    {
        match self.data.compare_exchange(
            current.data,
            new.data,
            ord.success(),
            ord.failure(),
        ) {
            Ok(_) => {
                let data = new.data;
                mem::forget(new);
                Ok(Ptr::from_data(data))
            }
            Err(previous) => Err((Ptr::from_data(previous), new)),
        }
    }
}
```
</details><br>

It's interesting that `swap_owned` is missing, although it is not stricly speaking necessary: `a.swap_owned(o, ord, scope)` would be equivalent to `a.swap(o.into_ptr(scope), ord, scope)`. But, for the sake of consistency and convenience, we should probably have `swap_owned` as well.

There's a lot of duplication. We have one set of functions for `Ptr` and another for `Owned`. Let's try to unify them. First, we could introduce a trait that is implemented by `Ptr` and `Owned`:

```rust
/// A trait implemented by `Owned<T>` and `Ptr<T>`.
pub trait OwnedOrPtr<T> {
    fn into_data(self) -> usize;
    unsafe fn from_data(data: usize) -> Self;
}

impl<T> OwnedOrPtr<T> for Owned<T> {
    fn into_data(self) -> usize {
        let data = self.data;
        mem::forget(self);
        data
    }

    unsafe fn from_data(data: usize) -> Self {
        Owned::from_data(data)
    }
}

impl<'scope, T> OwnedOrPtr<T> for Ptr<'scope, T> {
    fn into_data(self) -> usize {
        self.data
    }

    unsafe fn from_data(data: usize) -> Self {
        Ptr::from_data(data)
    }
}
```

Second, let's clarify the return value of `compare_and_set_owned`. Instead of returning 
```rust
Result<Ptr<'scope, T>, (Ptr<'scope, T>, Owned<T>)>
```
we could return something like
```rust
Result<Ptr<'scope, T>, CompareAndSetError<'scope, T>>
```

But let's make the error type a little bit more generic:

```rust
pub struct CompareAndSetError<'scope, T: 'scope, N> {
    /// The previous value in the `Atomic<T>`.
    pub previous: Ptr<'scope, T>,
    /// The new value that the compare-and-set operation failed to write.
    pub new: N,
}
```

Finally, now we're ready to eliminate all those `*_owned` functions. The new set of store/swap/compare-and-set functions is much smaller than the one at the beginning of this comment.

```rust
impl<T> Atomic<T> {
    pub fn store<N>(&self, new: N, ord: Ordering)
    where
        N: OwnedOrPtr<T>,
    {
        self.data.store(new.into_data(), ord);
    }

    pub fn swap<'scope, N>(&self, new: N, ord: Ordering, _: &'scope Scope) -> Ptr<'scope, T>
    where
        N: OwnedOrPtr<T>,
    {
        Ptr::from_data(self.data.swap(new.into_data(), ord))
    }

    pub fn compare_and_set<'scope, N, O>(
        &self,
        current: Ptr<T>,
        new: N,
        ord: O,
        _: &'scope Scope,
    ) -> Result<Ptr<'scope, T>, CompareAndSetError<'scope, T, N>>
    where
        N: OwnedOrPtr<T>,
        O: CompareAndSetOrdering,
    {
        let new_data = new.into_data();
        match self.data.compare_exchange(
            current.data,
            new_data,
            ord.success(),
            ord.failure(),
        ) {
            Ok(_) => Ok(Ptr::from_data(new_data)),
            Err(previous) => Err(CompareAndSetError {
                previous: Ptr::from_data(previous),
                new: unsafe { N::from_data(new_data) },
            })
        }
    }
}
```

The obvious downside is that the `OwnedOrPtr<T>` trait makes function signatures slightly more complicated. But maybe it's not too bad...? Maybe it's worth it in the end? What do you think?

### PR DESCRIPTION
**Pull request by jeehoonkang**
Closes #30

### CODE REVIEW

File: src/atomic.rs [hunk 1]
```diff
@@ -396,64 +288,63 @@ impl<T> Atomic<T> {
...
-        new: Owned<T>,
+        current: Shared<T>,
+        new: P,
         ord: O,
         _: &'g Guard,
-    ) -> Result<Ptr<'g, T>, (Ptr<'g, T>, Owned<T>)>
+    ) -> Result<Shared<'g, T>, (Shared<'g, T>, P)>
     where
         O: CompareAndSetOrdering,
+        P: Pointer<T>,
```
Comment (ghost) [inline src/atomic.rs:310]: This looks kinda messy at the moment, but with in-band lifetimes (see [here](https://github.com/rust-lang/rust/pull/46051)) and impl Trait in argument position (see [here](https://github.com/rust-lang/rust/pull/45918)), this signature will be simplified to:

```rust
pub fn compare_and_set<P: Pointer<T>>(
    &self,
    current: Shared<T>,
    new: P,
    ord: impl CompareAndSetOrdering,
    _: &'g Guard,
) -> Result<Shared<'g, T>, (Shared<'g, T>, P)>
```

**Review by ghost — APPROVED** [review]
Nice! I really like how this panned out. :)

File: src/atomic.rs (hunk 1 above)
Comment (Vtec234) [reply src/atomic.rs:310]: The return type stands out as really sigil-heavy and could potentially be changed to `Result<Shared<'g, T>, CasError<'g, T, P>>` with fields like `curr` and `new` in `CasError` or even entirely to `CasResult<'g, T, P>`. That said, the current interface uses standard types and is understandable without looking up more custom types, so i'm not so sure. The linked RFCs will definitely improve things.

**Review by Vtec234 — APPROVED** [review]

File: src/atomic.rs (hunk 1 above)
Comment (ghost) [reply src/atomic.rs:310]: I'm in favor of `CompareAndSetError<'g, T, P>` because it's easy to mistake the tuple `(previous, new)` for `(new, previous)`.

**Pull request merged**